### ISSUE TITLE
vkCreateDevice doesn't check features against enabled extensions

### ISSUE THREAD

**Issue by sfricke-samsung** [KhronosGroup/Vulkan-Docs#1534]
Moved from https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/2818

To sum up the issue:

```cpp
VkPhysicalDevice16BitStorageFeaturesKHR ext_feature = {};

VkPhysicalDeviceFeatures2 physical_features2 = {};
physical_features2.pNext = &ext_feature;

// Assume VK_KHR_get_physical_device_properties2 enabled at instance level
vkGetPhysicalDeviceFeatures2(physical_device, &physical_features2);

assert(ext_feature.storagePushConstant16 == VK_TRUE);

VkDeviceCreateInfo device_info = {};
device_info.pNext = &physical_features2;
device_info.enabledExtensionCount = 0; // VK_KHR_16bit_storage NOT enabled
vkCreateDevice(&device_info);
```

In the case of a 1.0 Vulkan implementation, there seems to be no explicit VU saying

> If you enable the supported feature, the device version must support it OR the device extension must enable it

@rua pointed to spots in the spec, but currently the Validation Layers are not checking for this

**Comment by krOoze** [comment]
I believe if device does not support a feature, it would just report the feature as unsupported.

If the feature requires an extension, then, well, you cannot generally use structs of extension that is not enabled:

> Each structure present in the `pNext` chain **must** be defined at runtime by either:
> - a core version which is supported
> - an extension which is enabled
> - a supported device extension in the case of physical-device-level functionality added by the device extension

**Comment by r-potter** [comment]
I think this is slightly more subtle, since this is a weird corner of the spec where we mandate one thing for applications and then state what implementations are supposed to do with unrecognised structures. My understanding here is that:

* If the device supports an core version/extension that provides a feature, then the extension **must** be enabled before the feature can be used in `vkCreateDevice`
* However, if the device doesn't support the necessary extension, then the skipping unrecognised structures clause is supposed to apply. In this case, `vkGetPhysicalDeviceFeatures2` should not populate the struct and `vkCreateDevice` should not attempt to enable features **regardless of the contents of the structure** (because an implementation logically cannot enable features it doesn't comprehend). So what happens when you attempt to enable a feature where the extension is not present at all is also well defined (i.e. nothing)

> Any component of the implementation (the loader, any enabled layers, and drivers) must skip over, without processing (other than reading the sType and pNext members) any extending structures in the chain not defined by core versions or extensions supported by that component. 

@Tobski I didn't explain myself well in todays call. Does the above interpretation match your understanding?

**Comment by krOoze** [comment]
> If the device supports an core version/extension that provides a feature, then the extension must be enabled before the feature can be used in `vkCreateDevice`

How would you enable the feature in the first place, if the extension is not enabled. It would be invalid to link the feature enabling struct in `vkCreateDevice`...

> However, if the device doesn't support the expose extension, then the skipping unrecognised structures clause is supposed to apply. In this case,

That feels somewhat irrelevant, because you are already in undefined behavior land.

I think those two clauses can coexist. Loader can add its own internal structures, which the implementation must ignore. And the app is not allowed to chain unknown structures.

**Comment by r-potter** [comment]
The reasoning behind this logic was that if a layer implements an extension, and a driver underneath it does not, then an application should be able to make use of that extension, and the layer should not have to rebuild the pNext chain (removing the unrecognised structure) before passing it down to the driver/layers below. 

Consequently we end up in a situation where:

* Applications should not use features without enabling the relevant extension
* Drivers and layers still have to behave sanely, and are not allowed to just crash, when passed an unrecognised structure...
* ...but may invoke undefined behaviour when passed a structure that they recognise (i.e. implement the relevant extension), but for which the necessary extension has not been enabled

Previous internal discussion here (for anyone following along with access): https://gitlab.khronos.org/vulkan/vulkan/-/issues/2442

**Comment by Rua** [comment]
Perhaps slightly off-topic, but if the application is not supposed to use anything the driver doesn't support, how does that work in the specific case of retrieving the device properties? One would use `vkGetPhysicalDeviceProperties(2)` to retrieve them, but you must not include any structures in the chain if they aren't supported. Yet one of the things that this function retrieves is the `apiVersion`, which is exactly what tells you what is supported. So you get a kind of chicken-and-egg problem, which seems only solvable if you call the function twice: first with no chained structures to retrieve the `apiVersion`, then a second time with the chain including whatever is supported by that `apiVersion`. Is this actually the intended usage?

**Comment by krOoze** [comment]
@r-potter I believe you just reworded what I said, though you did a pretty good job of it there! Did we also reach the same conclusion that there is not an issue and the feature VUs are not missing, because it is invalid to link the structures containing them in the first place?

@Rua I quoted the relevant parts above, specifically:

> Each structure present in the `pNext` chain **must** be defined at runtime by either:
> [...]
>  - a supported device extension in the case of physical-device-level functionality added by the device extension

there's also this:

> Physical-device-level functionality or behavior added by a device extension to the API **must** not be used unless the conditions described in Extending Physical Device Core Functionality are met.

I.e. you can use the versions and extensions in pre-device commands.

There's a subtle spec bug in there. `vkCreateDevice` could be considered a physical-device-level command, but I am pretty sure it isn't for this purpose. I.e. if you use some extension in the command (e.g. `pNext` struct), it also have to be being enabled in its `ppEnabledExtensionNames`.

There's no chicken-egg problem, because `vkGetPhysicalDeviceProperties2KHR` comes from an instance extension. Though I think there's another spec bug there (#issue_number_placeholder).

**Comment by Rua** [comment]
> I think there's no chicken-egg problem, because `vkGetPhysicalDeviceProperties2KHR` comes from an instance extension. Though I think there's spec bug there which would make you think that (#issue_number_placeholder).

I'm not talking about that function itself, that can be assumed to be present or not based on the instance. The question is whether the `pNext` chain of `VkPhysicalDeviceProperties2(KHR)` is allowed to contain structures that belong to e.g. Vulkan 1.2 if the _physical device_ itself supports only Vulkan 1.1. Because if not, which the spec seems to say, then in order to not violate the spec, the application must first call `vkGetPhysicalDeviceProperties2KHR` with no chain, in order to know the device's supported version, and only then include the `pNext` chain and call it a second time.

**Comment by krOoze** [comment]
@Rua Yes I understood. Though the command itself has the same problem because of what the spec says. As I recall the layers use it the "sane" way though. Anyway yea, let's make a new Issue for it. Do you want the honors, or should I?

**Comment by Rua** [comment]
If you want to, please. I'm not very familiar with things here yet.
